src/nvme/nbft.c

```diff
@@ -0,0 +1,947 @@
+
+/*
+ * HEADER (Figure 8)
+ */
+struct __attribute__((__packed__)) raw_nbft_header {
+	char signature[4];
+	__u32 length;
+	__u8 major_revision;
+	__u8 checksum;
+	char oem_id[6];
+	char oem_table_id[8];
+	__u32 oem_revision;
+	__u32 creator_id;
+	__u32 creator_revision;
+	__u32 heap_offset;
+	__u32 heap_length;
+	heap_obj driver_dev_path_sig;
+	__u8 minor_revision;
+	__u8 reserved[13];
+};
+
+/*
+ * CONTROL (Figure 8)
+ */
+struct __attribute__((__packed__)) raw_nbft_control {
+	__u8 structure_id;
+	__u8 major_revision;
+	__u8 minor_revision;
+	__u8 reserved1;
+	__u16 length;
+	__u8 flags;
+	__u8 reserved2;
+
+	heap_obj host_descriptor;
+	__u8 host_descriptor_version;
+	__u8 reserved3;
+
+	__u32 hfi_descriptor_list_offset;
+	__u16 hfi_descriptor_length;
+	__u8 hfi_descriptor_version;
+	__u8 num_hfi;
+
+	__u32 ssns_descriptor_list_offset;
+	__u16 ssns_descriptor_length;
+	__u8 ssns_descriptor_version;
+	__u8 num_ssns;
+
+	__u32 security_profile_descriptor_list_offset;
+	__u16 security_profile_descriptor_length;
+	__u8 security_descriptor_version;
+	__u8 num_sec;
+
+	__u32 discovery_profile_descriptor_list_offset;
+	__u16 discovery_profile_descriptor_length;
+	__u8 descovery_descriptor_version;
+	__u8 num_disc;
+
+	__u8 reserved4[16];
+};
+
+#define CONTROLFLAG_VALID		0x01
+
+/*
+ * HOST DESCRIPTOR (Figure 9)
+ */
+struct __attribute__((__packed__)) raw_nbft_host {
+	__u8 structure_id;
+	__u8 flags;
+	uuid_t host_identifier;
+	heap_obj host_nqn;
+	__u8 reserved[8];
+};
+
+#define HOSTFLAG_VALID			0x01
+#define HOSTFLAG_HOSTID_CONFIGURED	0x02
+#define HOSTFLAG_HOSTNQN_CONFIGURED	0x04
+#define HOSTFLAG_PRIMARY_ADMIN_HOST	0x18
+
+enum nbft_transport_types {
+	nbft_trtype_tcp = 3,
+};
+
+/*
+ * HFI DESCRIPTOR (Figure 11)
+ */
+struct __attribute__((__packed__)) raw_nbft_hfi {
+	__u8 structure_id;
+	__u8 index;
+	__u8 hfi_flags;
+	__u8 hfi_transport_type;
+	__u8 reserved1[12];
+	heap_obj hfi_transport_descriptor;
+	__u8 reserved2[10];
+};
+
+#define HFIFLAG_VALID		0x01
+
+/*
+ * HFI TRANSPORT INFO DESCRIPTOR (Figure 13)
+ */
+struct __attribute__((__packed__)) raw_nbft_hfi_info_tcp {
+	__u8 structure_id;
+	__u8 version;
+	__u8 hfi_transport_type;
+	__u8 transport_info_version;
+	__u16 hfi_index;
+	__u8 transport_flags;
+	__u32 pci_sbdf;
+	__u8 mac_addr[6];
+	__u16 vlan;
+	__u8 ip_origin;
+	__u8 ip_address[16];
+	__u8 subnet_mask_prefix;
+	__u8 ip_gateway[16];
+	__u8 reserved1;
+	__u16 route_metric;
+	__u8 primary_dns[16];
+	__u8 secondary_dns[16];
+	__u8 dhcp_server[16];
+	heap_obj host_name;
+	__u8 reserved2[18];
+};
+
+#define HFIINFOTCPFLAG_VALID		0x01
+#define HFIINFOTCPFLAG_GLOBAL_ROUTE	0x02
+#define HFIINFOTCPFLAG_DHCP_OVERRIDE	0x04
+
+/*
+ * SUBSYSTEM NAMESPACE DESCRIPTOR (Figure 15)
+ */
+struct __attribute__((__packed__)) raw_nbft_ssns {
+	__u8 structure_id;
+	__u16 index;
+	__u16 flags;
+	__u8 transport_type;
+	__u16 transport_specific_flags;
+	__u8 primary_discovery_ctrl_index;
+	__u8 reserved1;
+	heap_obj subsystem_transport_address;
```

Comment:
Just nitpicking - The naming convention for "transport address" is `traddr` and for "transport service id" is `trsvcid`.

This comment applies to both `nbft.c` and `nbft.h`. External code such as `nvme-cli` that refers to these definitions will obviously be affected as well.